File: backend/app/core/password_policy.py
from __future__ import annotations

import re

# Contraseñas demasiado comunes (subset; ampliar según necesidad).
COMMON_PASSWORDS = frozenset(
    {
        "password",
        "password1",
        "password123",
        "12345678",
        "123456789",
        "1234567890",
        "qwerty123",
        "admin123",
        "biblioteca",
        "centro123",
    }
)


def validate_password_strength(password: str) -> str | None:
    """Devuelve mensaje de error o None si la contraseña es válida."""
    if len(password) < 10:
        return "La contraseña debe tener al menos 10 caracteres."
    if len(password) > 128:
        return "La contraseña no puede superar 128 caracteres."
    if not re.search(r"[a-z]", password):
        return "Incluí al menos una letra minúscula."
    if not re.search(r"[A-Z]", password):
        return "Incluí al menos una letra mayúscula."
    if not re.search(r"\d", password):
        return "Incluí al menos un número."
    if password.lower() in COMMON_PASSWORDS:
        return "Esa contraseña es demasiado común. Elegí otra."
    return None

File: frontend/lib/password-policy.ts
import { z } from 'zod'

const COMMON = new Set([
  'password',
  'password1',
  'password123',
  '12345678',
  '123456789',
  '1234567890',
  'qwerty123',
  'admin123',
  'biblioteca',
  'centro123',
])

export function passwordStrengthMessage(password: string): string | null {
  if (password.length < 10) return 'Mínimo 10 caracteres.'
  if (password.length > 128) return 'Máximo 128 caracteres.'
  if (!/[a-z]/.test(password)) return 'Incluí una minúscula.'
  if (!/[A-Z]/.test(password)) return 'Incluí una mayúscula.'
  if (!/\d/.test(password)) return 'Incluí un número.'
  if (COMMON.has(password.toLowerCase())) return 'Contraseña demasiado común.'
  return null
}

export const registerPasswordSchema = z
  .string()
  .min(10, 'Mínimo 10 caracteres')
  .max(128)
  .refine((p) => /[a-z]/.test(p), 'Incluí una minúscula')
  .refine((p) => /[A-Z]/.test(p), 'Incluí una mayúscula')
  .refine((p) => /\d/.test(p), 'Incluí un número')
  .refine((p) => !COMMON.has(p.toLowerCase()), 'Contraseña demasiado común')

export function passwordScore(password: string): number {
  let score = 0
  if (password.length >= 10) score += 1
  if (password.length >= 14) score += 1
  if (/[a-z]/.test(password) && /[A-Z]/.test(password)) score += 1
  if (/\d/.test(password)) score += 1
  if (/[^A-Za-z0-9]/.test(password)) score += 1
  if (COMMON.has(password.toLowerCase())) score = 0
  return Math.min(score, 4)
}
